Allow Admin to have multiple parents

robhunt3r opened the pull request:
I've found that our project required Admin to have multiple parent Admins, but sonata couldn't handle that, as it was only possible to have one parent at a time.

With this, we allow an Admin to use multiple parents, it is working in our project, I've changed the tests file but i don't know if I've edited good, because I don't know about theese Tests...

How does it works?

Well, `$parentAssociationMapping` is now an Array which contains a key based on an admin's code and a value which is the entity mapped field.

For example

`Post` has `Comments`
`Page` has `Comments`
`Comment` can be from `Post` or `Page`.

`Post` = `admin.post`
`Page` = `admin.page`
`Comment` = `admin.comment`

Our Yaml file would be

``` yaml
  admin.post:
      class: AdminBundle\Admin\Post
      arguments: [~, AdminBundle\Entity\Post, ~]
      tags:
          - {name: sonata.admin, manager_type: orm, group: admin, label: post}
      calls:
          - [addChild, [@admin.comment, "post"]]

  admin.page:
      class: AdminBundle\Admin\Page
      arguments: [~, AdminBundle\Entity\Page, ~]
      tags:
          - {name: sonata.admin, manager_type: orm, group: admin, label: page}
      calls:
          - [addChild, [@admin.comment, "page"]]

  admin.comment:
      class: AdminBundle\Admin\Comment
      arguments: [~, AdminBundle\Entity\Comment, ~]
      tags:
          - {name: sonata.admin, manager_type: orm, group: admin, label: comment}
```

As you can see, instead of declaring `$parentAssociationMapping` field inside `Admin` class, we declare the linked field inside our admin's config, this would add an array to `Comment` that would be

``` php
array('admin.post' => 'post', 'admin.page' => 'page');
```

`post` and `page` are `manyToOne` relationships between `Comment` and `Post`/`Page` so, it's the field name as always have been used in Sonata.

It will call `setParent()` as usual. So when retrieving `getParentAssociationMapping()` we check if we have a parent, if we do, we will get it's admin's code, and check against the array, return the value (the field) or throw an exception if the key was not found.

Maybe it can be improved, but I think this is something needed and I've searched it everywhere but people was doing ugly things like getting the class name on a switch-case that of course is not automatic and need writing on every admin and every situation.

OskarStark commented on Admin/Admin.php:
```diff
@@ -287,9 +287,9 @@
+     * @var array
      */
-    protected $parentAssociationMapping = null;
+    protected $parentAssociationMapping = array();
```
This is BC break

OskarStark commented on Admin/Admin.php:
```diff
@@ -1956,12 +1979,13 @@ public function getFilterFieldDescriptions()
      * {@inheritdoc}
      */
-    public function addChild(AdminInterface $child)
+    public function addChild(AdminInterface $child, $field)
```
This is BC break

OskarStark commented on Admin/AdminInterface.php:
```diff
@@ -520,11 +520,12 @@ public function getLabelTranslatorStrategy();
+     * @param string $field
      */
-    public function addChild(AdminInterface $child);
+    public function addChild(AdminInterface $child, $field);
```
This is BC break

robhunt3r commented:
@OskarStark You're totally right, I think I've fixed it. I've pushed to that branch, how can I update the PR with this new commit thas fixes BC?

OskarStark commented:
pleas fix the CS issues from travis:
```
Legend: ?-unknown, I-invalid file syntax, file ignored, .-no changes, F-fixed, E-error
   1) Admin/Admin.php (phpdoc_separation, return, braces)
   2) Admin/AdminInterface.php (phpdoc_short_description, phpdoc_params, eof_ending)
```

robhunt3r commented:
@OskarStark I really don't know what you are talking about... New to travis.

OskarStark commented:
we use some coding standards an travis check these standard and i gave you the output.

after that please squash your commits.
@see https://github.com/sonata-project/SonataAdminBundle/blob/master/CONTRIBUTING.md#squash-your-commits

OskarStark commented on Admin/Admin.php:
```diff
@@ -918,11 +918,35 @@ public function buildDatagrid()
      * value (ie the parent object) or to filter the object.
      *
-     * @return string the name of the parent related field
+     * @return null|string
```
please add a blank line after ``@return null|string``

OskarStark commented on Admin/Admin.php:
```diff
@@ -918,11 +918,35 @@ public function buildDatagrid()
+
+    /**
+     * @param string $code
+     * @param string $value
```
please add a blank line after ``@param string $value``

OskarStark commented on Admin/Admin.php:
```diff
@@ -918,11 +918,35 @@ public function buildDatagrid()
+    /**
+     * @param string $code
+     * @param string $value
+     * @return $this
```
the method does not return something

OskarStark commented on Admin/AdminInterface.php:
```diff
@@ -520,11 +520,12 @@ public function getLabelTranslatorStrategy();
+     * add an Admin child to the current one linked to child's field
      *
      * @param \Sonata\AdminBundle\Admin\AdminInterface $child
+     * @param null $field
```
``null|string`` i think

OskarStark commented on Admin/AdminInterface.php:
```diff
@@ -1019,4 +1020,4 @@ public function getAccessMapping();
      */
     public function checkAccess($action, $object = null);
-}
+}
```
no new line @ end of file

OskarStark commented:
i added some comments for the CS problems

robhunt3r commented:
@OskarStark Okay thanks. I've run `make cs` and rebased I hope everything it's correct now.

OskarStark commented:
@rande can you please verify this PR?

Koc commented:
@robhunt3r can you add docs that described this new feature, please?

robhunt3r commented:
@Koc Done, I think it's clear enough, if not just tell me and I will update it again.

Please note that I've removed ``$parentAssociationMapping`` from docs because now it is automatically updated with an array of parent admins so it is not needed anymore.

OskarStark commented:
can you please rebase @robhunt3r

robhunt3r commented:
@OskarStark It should be git rebase upstream/master, right?

robhunt3r commented:
Well, I forced the push and seemed to work @OskarStark . Dunno what happened

OskarStark commented:
seems to be correct

can you please fix the CS issue?

``make cs``

Thank you @robhunt3r

robhunt3r commented:
@rande @OskarStark Any news about merging this? We are using it but we would like not to use our fork, it's a headache when whe need to pull from sonata master and rebase this PR...

core23 commented on Admin/Admin.php:
```diff
@@ -923,11 +923,37 @@ public function buildDatagrid()
+     */
+    public function addParentAssociationMapping($code, $value)
+    {
+        $this->parentAssociationMapping[$code] = $value;
```
``parentAssociationMapping`` could be null

greg0ire replied:
@robhunt3r, can you fix that quickly ? We are trying to merge all [these PRs](https://github.com/sonata-project/SonataAdminBundle/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aopen+milestone%3A2.4+) before doing some big changes on the project.

robhunt3r replied:
I am not available to fix this until Monday, can you wait? I will do first time in the morning as soon as I go to my office.

Enviado desde mi iPhone

> El 9 abr 2016, a las 14:51, Grégoire Paris <notifications@github.com> escribió:
> 
> In Admin/Admin.php:
> 
> > +                ));
> > +            }
> > +        } else {
> > +            return $this->parentAssociationMapping;
> > +        }
> > +    }
> > +
> > +    /**
> > +     * @param string $code
> > +     * @param string $value
> > +     *
> > +     * @return $this
> > +     */
> > +    public function addParentAssociationMapping($code, $value)
> > +    {
> > +        $this->parentAssociationMapping[$code] = $value;
> @robhunt3r, can you fix that quickly ? We are trying to merge all these PRs before doing some big changes on the project.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

greg0ire replied:
I think I can fix this myself, so I'll will rather do that.

greg0ire replied:
Wait, actually, this does not seem to be a problem for php @core23 : 

```
php -a
Interactive shell

php > $a = null;
php > $a['test'] = 'pwet';
php > var_dump($a);
array(1) {
  ["test"]=>
  string(4) "pwet"
}
```

core23 replied:
I'm not 100% sure, but some php versions have problems with this.

greg0ire replied:
Let's be sure then : https://3v4l.org/VtugS

core23 replied:
Ignore my comment ;)

greg0ire replied:
Great! Can this be merged then?

ju1ius replied:
@core23 @greg0ire This behaviour has been here for ages. It even works with nested keys.
It is however explicitely discouraged in the [documentation](http://fr2.php.net/manual/en/language.types.array.php#language.types.array.syntax.modifying). Too much wtf factor.

greg0ire replied:
Ok

core23 commented on Admin/Admin.php:
```diff
@@ -923,11 +923,37 @@ public function buildDatagrid()
+     * @param string $code
+     * @param string $value
+     *
+     * @return $this
```
Remove the ``@return`` annoation or add it to the code

core23 commented on Admin/Admin.php:
```diff
@@ -292,7 +292,7 @@
      * the order (OrderElement::$order).
      *
-     * @var \ReflectionProperty
+     * @var mixed
```
Can we specify this? 

``\ReflectionProperty|something else``

greg0ire replied:
Uuuuuh… where does this reflection property stuff come from? Isn't this variable supposed to be a string?

greg0ire replied:
Looked at the history, and it seems that the comment is outdated since b0df44f9894b08ea1ac526231990547919da375f

ju1ius commented on Admin/Admin.php:
```diff
@@ -923,11 +923,37 @@ public function buildDatagrid()
+            }
+        } else {
+            return $this->parentAssociationMapping;
+        }
```
a return in else clause ?
isn't it a cs issue ?

greg0ire replied:
It would be reported by `phpmd` I think. The else should be removed.

greg0ire commented:
Fixed this in #3711

greg0ire commented on Resources/doc/reference/architecture.rst:
```diff
@@ -330,30 +332,10 @@ To do this, you first need to call the ``addChild`` method in your PostAdmin ser
 
             <call method="addChild">
                 <argument type="service" id="sonata.news.admin.comment" />
+                <argument>post</<argument>
```
This line is not valid xml, I also fixed that in my PR.

greg0ire commented:
I think this can't be fixed soon enough, and should be removed from the 2.4 milestone (see my alternate PR, which I closed).

robhunt3r commented:
I've fixed notes you told me.

About the array being null, well, it's been there for years in PHP, never gave me an error on any script and it's the best workaround to let the property be an Array and a String at the same time, depending of it's contents.

greg0ire commented:
Please run `make cs` to fix the build.

greg0ire commented:
And what about the `getParent()` method? Does it still make sense?

robhunt3r commented:
`make cs` done.
If we remove `getParent()` method, wont it bc break?

greg0ire commented:
>  If we remove getParent() method, wont it bc break?

Of course it will. Maybe it should return the first parent ? Or the "current" parent? Please look at it, see what it is used for, and decide what should be done.

robhunt3r commented:
Well, it returns the using parent, in the example, if we are viewing comments from `Post`, parent will be `Post`... At least that's how it is working in our project.

greg0ire commented:
> Well, it returns the using parent, in the example, if we are viewing comments from Post, parent will be Post... At least that's how it is working in our project.

Really? Here is the code for `addChild()` : 

```
$this->children[$child->getCode()] = $child;

$child->setBaseCodeRoute($this->getCode().'|'.$child->getCode());
$child->setParent($this);
```

IMO, the last to call `addChild` will get the "child custody" :wink: . Can you add a `dump(__CLASS__);` inside this method, and see if it somehow is called in a different order when you look at the admin contextualized under each parent?

core23 commented on Admin/Admin.php:
```diff
@@ -923,7 +923,7 @@ public function buildDatagrid()
         $this->datagrid->getPager()->setMaxPageLinks($this->maxPageLinks);
 
-        $mapper = new DatagridMapper($this->getDatagridBuilder(), $this->datagrid, $this);
+        $mapper = new DatagridMapper($this->getDatagridBugetParentilder(), $this->datagrid, $this);
```
getDatagridBugetParen**tilder**? Looks like a few chars are missing @robhunt3r

core23 commented on Admin/Admin.php:
```diff
@@ -2030,12 +2057,15 @@ public function getFilterFieldDescriptions()
      * {@inheritdoc}
      */
-    public function addChild(AdminInterface $child)
+    public function addChild(AdminInterface $child, $field = null)
```
This is a BC break

Koc replied:
Why? This isn't interface change.

core23 replied:
Also a public method was changed, which is a break

greg0ire replied:
@Koc : It's not a BC-break for calling code, but it is a BC break for extending code. [See for yourself](https://3v4l.org/PDY3L)

OskarStark replied:
so if we cannot handle this, we need to move this to another milestone

greg0ire replied:
@OskarStark done

OskarStark replied:
thank you @greg0ire

core23 commented on Admin/AdminInterface.php:
```diff
@@ -528,11 +528,12 @@ public function getLabelTranslatorStrategy();
+     * @param null|string    $field
      */
-    public function addChild(AdminInterface $child);
+    public function addChild(AdminInterface $child, $field = null);
```
The interface changes here @Koc

Koc replied:
I haven't changed this interface)

Soullivaneuh commented:
Why major @greg0ire Can't this be BC?

greg0ire commented:
I tagged this as it is, if this is made BC, the tag can be changed.

Soullivaneuh commented:
> if this is made BC, the tag can be changed.

@robhunt3r If you can make it BC, please reopen the PR on the 3.x branch.

robhunt3r commented:
I will take a look as soon as I can. 

As @greg0ire pointed, it is a BC-break for extending, not for calling so using it in 3.x branch wont be BC-break as it's a major branch?

And as @core23 says, `getDatagridBugetParentilder` it's an epic fail, PHPStorm sometimes makes stuff like this without a reason... I will fix that too.

greg0ire commented:
>  it's a major branch

There is no such thing as a major branch, only major changes, and branches for the next major release. A BC-break is only defined between two commits. It's implicitly between your PR and the last release (so 3.0.0)

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

robhunt3r commented:
@SonataCI done, with master branch, must I rebase on 3.x branch?

Soullivaneuh commented:
@robhunt3r https://github.com/sonata-project/SonataAdminBundle/pull/3339#issuecomment-217747600

Please not that @SonataCI is just a bot posting message on conflicted PR. :wink:

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

core23 commented:
Please rebase your PR and use the new PR template @robhunt3r

core23 commented:
Ping @robhunt3r

core23 commented:
Please rebase @robhunt3r

core23 commented:
Are you still working on this @robhunt3r ? Would appreciate this feature ❤️

core23 commented on Admin/AbstractAdmin.php:
```diff
@@ -822,14 +822,41 @@ public function buildDatagrid()
+     *
+     * @return $this
+     */
+    public function addParentAssociationMapping($code, $value)
```
Please make this final

haivala commented:
Unbelievable that this hasn't find it's way to the code

OskarStark commented:
@haivala you can take care of this PR and apply the comments we made, if you like

greg0ire commented:
And if you can't find the courage to do so, then maybe you will find it less unbelievable.

robhunt3r commented:
Sorry guys, I didn't notice this things, too much time.

I can't remember how to do the squash + rebase from remote branch, if you can give me the tips I'll be more than happy to end this PR.

greg0ire commented:
1. Fetch all new commits: `git fetch --all`. ⬇️ 
2. Rebase on what you fetched interactively: ✂️ 
    1. `git rebase origin/3.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin repository.
   2. If you run into a conflict: 💥 
        1. fix it with `git mergetool`. 😷 
        1. continue on your merry way with `git rebase --continue`. ⏩ 
3. Force push to overwrite all this : `git push --force`. ⬆️

robhunt3r commented:
Thanks @greg0ire but I am getting a ton of conflicts (even tho I only changed 4 files...) a lot of them are just formatting conflicts, but in files that I've never changed... these are the steps I've followed in order to rebase and squash

```
git clone https://github.com/robhunt3r/SonataAdminBundle
cd SonataAdminBundle/
git remote add upstream https://github.com/sonata-project/SonataAdminBundle
git fetch -all
git checkout multiple_admin_parent
git rebase upstream/3.x
```
As I said, at a point I got an infinite list of conflicts, it will take me ages to make sure my changes are there... Am I doing something wrong? Would it better to just clone the 3.x branch and do my changes on that and closing this PR?

Also, I remember that `make cs` fixed some formatting, but it isn't working anymore... It's been a long since I do PR on Symfony bundles ^^'

greg0ire commented:
Should be `git rebase -i upstream/3.x`, then you remove all the lines you do not recognized as created by you, you will get a lot less conflicts.

robhunt3r commented:
Thanks for the fast reply @greg0ire 

When I do that (if you see, I've rebased my 2 commits for simplicity) it shows me this lines

```
  GNU nano 2.5.3                                                         Archivo: /home/moises/Dev/SonataAdminBundle/.git/rebase-merge/git-rebase-todo                                                                                                                          

pick 20bb1bd 4.x-dev
pick 249dcbf 2.x-dev
pick c26ac43 DevKit updates
pick 8cbb890 DevKit updates
pick ca52f41 DevKit updates (#3775)
pick 3d6e7f4 Authorize block-bundle ^4.0@dev (#3777)
pick d637f0f DevKit updates for master branch (#3778)
pick e5fab51 DevKit updates
pick e127250 Apply CS
pick 0e4d124 Add configureActionButtons to the admin interfaces (#3606)
pick 041097d Added "configureBatchActions" method to AdminExtensionInterface (#3550)
pick ef288a6 DevKit updates for master branch (#3801)
pick a5c06a5 4.x on changelog
pick 6ebbabe Remove deprecated Admin class (#3810)
pick ace4e92 DevKit updates (#3814)
pick 297dc9d DevKit updates for master branch (#3815)
pick f87519f Remove BC handler for deprecated 'view' _action (#3829)
pick 4337ac3 Added "getAccessMapping" method to AdminExtensionInterface (#3539)
pick 001d575 Remove default template loading on exception (#3704)
pick 5279066 Make the twig extension final
pick 3ec0b5c Make label translator strategies private
pick 3ff2dde Test methods without making them public
pick 8088d96 Admin can have multiple parents now
```

The last one is my commit, `8088d96`, for squashing I just prepend "f" or "s" (depends on the needs), as you told me if I remove all but my commit, and prepend "pick" when I go into my conflicts it says that I have conflicts for those 4 files (correct)

When I check them with git mergetool it shows me this, which is weird because how can it be 'deleted' ?
Example:
```
Deleted merge conflict for 'Admin/AbstractAdmin.php':
  {local}: deleted
  {remote}: modified file
Use (m)odified or (d)eleted file, or (a)bort? 
```
Sorry for the annoyance btw

greg0ire commented:
It's not weird, they were "deleted" indeed because they were in fact moved to `src`.

robhunt3r commented:
Oh I see it, true. 

So I guess I should keep my files deleted and apply the changes to the new ones? otherwise I won't be able to "merge" them, as they are different files (or at least, detected as that).

I will go with it and give feedback, damn, this is what happens when I don't get notified in more than a year about a PR -.-

robhunt3r commented:
@greg0ire Everytime I try to do it, my PR ends up showing commits from other people, not mine.
```
git clone https://github.com/robhunt3r/SonataAdminBundle
cd SonataAdminBundle/
git remote add upstream https://github.com/sonata-project/SonataAdminBundle
git fetch -all
git checkout multiple_admin_parent
git rebase -i upstream/3.x
```
With that, I've deleted all lines but my commit, showing me conflicts for those 4 files becuase they are "deleted" (they're at a new location) so I just accept upstream changes so my files just vanish so I can write my code in the new files, so everything can work fine, which is what I want, but then it says
```
Your branch and 'origin/multiple_admin_parent' have diverged,
and have 655 and 33 different commits each, respectively.
```
If I try to push I can't, if I do with a `--force` then other people commits shows up.

I've also tried with a merge, no diverged at any moment, everything went just fine, edited those 4 files at their new locations, commited and pushed.

When I push, again, commits from other people... 

I know this can sound stupid for all of you but I am used to use git in a more simple way in my work, so forgive me for being so lame ^^'


Anyway, I've reverted all changes to the point my PR was, just in case something went fucked up.

franmomu commented:
Hi! I was taking a look at this and I tried [this solution](https://stackoverflow.com/a/15137607) and apparently it worked. This is what I did (from my fork):

```
git clone https://github.com/robhunt3r/SonataAdminBundle
cd SonataAdminBundle/
git remote add upstream https://github.com/sonata-project/SonataAdminBundle
git fetch --all
git checkout multiple_admin_parent
git checkout -b temp
git reset --hard HEAD~1
mkdir -p src/Admin docs/reference
git mv Admin/AbstractAdmin.php src/Admin/AbstractAdmin.php
git mv Admin/AdminInterface.php  src/Admin/AdminInterface.php
git mv Resources/doc/reference/architecture.rst docs/reference/architecture.rst
git mv Tests/Admin/AdminTest.php tests/Admin/AdminTest.php
git commit -m 'rename'
git rebase temp multiple_admin_parent
git rebase temp multiple_admin_parent --onto bdb34ecfba7e8cef87978b2fac2ce6695193439a
# Then you have to fix the conflicts and git rebase --continue
git rebase -i upstream/3.x
# Fix conflicts and continue
```
The bdb34ecfba7e8cef87978b2fac2ce6695193439a commit is the one that changes the folder structure.

jordisala1991 commented:
Rebasing this old pull request could be a huge pain. @greg0ire won't approve, but if this is so hard to rebase. Just create a new pull request with the changes

greg0ire commented:
@franmomu you could also publish your changes and let @robhunt3r `reset --hard` to them :P Or carry the PR yourself (so, in a new PR)

franmomu commented:
Sorry! I was curious on how to solve git conflicts with renames and so on 😬and I thought it was interesting to share, I'm submitting the PR.

jordisala1991 commented:
Closing because #4954 is the same
